### PR TITLE
Procedure execution timeout

### PR DESCRIPTION
**Pull request by vara855**
- [x] tests and linter show no problems (`npm t`)
- [x] tests are added/updated for bug fixes and new features
- [x] code is properly formatted (`npm run fmt`)

Closes: #1481 

@tshemsedinov help me with next questions: 
* do i need to make timeoutFn as private class method, or move it somewhere to utils? 
* do i need to make a custom error class for timed out case? 

I'll make a changes from your suggestions. Have a nice day, this tech stack is awesome 👏

### CODE REVIEW

**Review by tshemsedinov — COMMENTED** [review]
We don't need wrapper `timeoutFn` in this case, better improve `metautil.timeout` to make it cancelable and use `Promise.race` with both promises.

**Comment by tshemsedinov** [comment]
I'd like to use [AbortController](https://nodejs.org/api/globals.html#globals_class_abortcontroller) for cancelable timeout but we have it only in node.js v15.0.0, so need simple polyfill (not full-featured), to change it in favour of `AbortController` later. I'll implement it in `metautil` ASAP.

**Comment by tshemsedinov** [comment]
@vara855 now we have cancelable timeout, so update metautil and use it. See implementation: https://github.com/metarhia/metautil/blob/5095b0d15c869241c3b37da23aebedbdf24c0735/lib/utilities.js#L145-L161

**Comment by vara855** [comment]
@tshemsedinov I find it strange that the timeout state is no result of `Promise.race`. Might the `matautil.timeout` should reject promise on setTimeout, not resolving?

**Comment by tshemsedinov** [comment]
@vara855 see metautil@3.5.0

File: lib/procedure.js [hunk 1]
```diff
@@ -1,6 +1,10 @@
 'use strict';
 
-const { Semaphore } = require('./dependencies.js').metarhia.metautil;
+const {
+  Semaphore,
+  timeout,
+  createAbortController,
+} = require('./dependencies.js').metarhia.metautil;
```
Comment (tshemsedinov) [inline lib/procedure.js:7]:
```suggestion
const { metautil } = require('./dependencies.js').metarhia;
const { Semaphore, createAbortController } = metautil;
```

File: lib/procedure.js [hunk 2]
```diff
@@ -40,7 +44,13 @@ class Procedure {
...
 
   async invoke(context, args = {}) {
-    const { script, parameters, validate, returns } = this;
+    const {
+      script,
+      parameters,
+      validate,
+      returns,
+      timeout: executionTimeout,
+    } = this;
```
Comment (tshemsedinov) [inline lib/procedure.js:53]:
```suggestion
    const { script, parameters, validate, returns, timeout } = this;
```

File: lib/procedure.js [hunk 3]
```diff
@@ -50,7 +60,17 @@ class Procedure {
     if (validate) {
       validate(args);
     }
-    const result = await method(args);
+    let result;
+    if (executionTimeout) {
```
Comment (tshemsedinov) [inline lib/procedure.js:64]:
```suggestion
    if (timeout) {
```

File: lib/procedure.js [hunk 4]
```diff
@@ -50,7 +60,17 @@ class Procedure {
...
+      const ac = createAbortController();
+      result = await Promise.race([
+        timeout(executionTimeout, ac.signal),
```
Comment (tshemsedinov) [inline lib/procedure.js:67]:
```suggestion
        metautil.timeout(timeout, ac.signal),
```

**Comment by tshemsedinov** [comment]
Landed in https://github.com/metarhia/impress/pull/1490/commits/03281db2ec9102260da5dca021352a7ada2e5596

**Comment by tshemsedinov** [comment]
@vara855 later we may profile this method and optimize it for performance but now it looks like working approach. Thanks!

**Pull request closed without merging**